pkg/workloads/console/controller.go

```diff
@@ -101,43 +108,80 @@ func (r *ConsoleReconciler) Reconcile(request reconcile.Request) (res reconcile.
 		return res, err
 	}
 
-	// Try to find the job for this console
+	// Try to find the job
 	job := &batchv1.Job{}
-	err = r.client.Get(r.ctx, request.NamespacedName, job)
+	err = r.client.Get(r.ctx, name, job)
 
 	// If it can't be found, create it
 	if err != nil {
-		if errors.IsNotFound(err) {
-			logger.Log("event", EventNotFound, "resource", Job)
-			err = r.client.Create(r.ctx, buildJob(request.NamespacedName, consoleTemplate.Spec.Template))
-			if err != nil {
-				return res, err
-			}
-			logger.Log(
-				"event", EventCreated,
-				"resource", Job,
-				"name", request.NamespacedName.Name,
-				"user", csl.Spec.User,
-			)
+		if !errors.IsNotFound(err) {
+			return res, err
 		}
 
-		return res, err
+		logger.Log("event", EventNotFound, "resource", Job)
+		job = buildJob(name, consoleTemplate.Spec.Template)
+		if err := controllerutil.SetControllerReference(csl, job, scheme.Scheme); err != nil {
+			return res, err
+		}
+		if err = r.client.Create(r.ctx, job); err != nil {
+			return res, err
+		}
+		logger.Log(
+			"event", EventCreated,
+			"resource", Job,
+			"name", job.ObjectMeta.Name,
+			"user", csl.Spec.User,
+		)
 	}
 
-	// If it exists:
-	//   Terminate if necessary
-	//   GC if necessary
 	logger.Log(
 		"event", EventFound,
 		"resource", Job,
-		"name", request.NamespacedName.Name,
+		"name", name.Name,
 		"user", csl.Spec.User,
 	)
 
+	// Find or create the role and role bindings
+	if err := r.updateRoleBindings(csl, consoleTemplate, name); err != nil {
+		return res, err
+	}
+
+	// TODO:
+	//   Terminate the console if it has expired
+	//   GC the terminated console if necessary
+
 	logger.Log("event", EventComplete)
 	return res, err
 }
 
+func (r *ConsoleReconciler) updateRoleBindings(csl *workloadsv1alpha1.Console, tmpl *workloadsv1alpha1.ConsoleTemplate, name types.NamespacedName) error {
+	role := buildRole(name)
+	if err := controllerutil.SetControllerReference(csl, role, scheme.Scheme); err != nil {
+		return err
+	}
+	if err := findOrCreate(r.ctx, r.client, role, name); err != nil {
```

Comment:
Yeah, I was aware of this when writing it. I think it would be nice to have a controller that will converge existing consoles to an updated state, as long as we can express that here in a nice way. I'd like to create a followup for this, as I hope the right approach will become clearer as we continue to flesh out the controller.

Reply:
Created a ticket to make sure we come back to this: https://gocardless.myjetbrains.com/youtrack/issue/PT-2485

Reply:
Great.